### ISSUE TITLE
Rule Proposal: prefer-destructuring

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#6053]
From [requireArrayDestructuring](http://jscs.info/rule/requireArrayDestructuring) and [requireObjectDestructuring](http://jscs.info/rule/requireObjectDestructuring).

This rule will warn `MemberExpression`s as the favor of destructuring assignments.

``` js
{
    "prefer-destructuring": ["error", {"object": true, "array": true}]
}
```

JSCS documents seem to say that this doesn't prefer nested destructuring.
How do we address nested destructuring?

**Comment by nzakas** [comment]
@markelog @mikesherov @mdevils @hzoo @zxqfox can you provide some insights for @mysticatea?

**Comment by zxqfox** [comment]
Nested destructures are not so easy to read as :
```js
var metadata = {
    title: "Scratchpad",
    translations: [
       {
           title: "JavaScript-Umgebung"
       }
    ]
};

var { title: englishTitle, translations: [{ title: localeTitle }] } = metadata;
// the same as
var englishTitle = metadata.title;
var localeTitle = metadata.translations[0].title;
```

Guess we just need to support the basic JSCS functionality to allow users to migrate.
But personally I'd say here should be an option to restrict or require nested ones.

**Comment by markelog** [comment]
> JSCS documents seem to say that this doesn't prefer nested destructuring.
How do we address nested destructuring?

Where?

Anyhow we can do this in two steps - first implement basic functionality and if there would be a request for it deal with nested stuff, I also think it could be an option disabled by default.

Also none of use actually did implementation for these rules.

@seanpdoyle Maybe you would like to help us port it these rules to ESLint?

**Comment by kaicataldo** [comment]
So it sounds like we have a few votes for feature parity with the JSCS equivalent rules with the possibility of adding nested checks in the future. Totally agree with @zxqfox and I think it makes sense to only go one level deep to start off with. Forcing nested destructuring seems like something most users wouldn't want on by default.

**Comment by platinumazure** [comment]
I think we can agree on one level for now, just so we have something for users.

All in favor? :+1: from me.

**Comment by nzakas** [comment]
We still need a champion to drive this forward.

**Comment by kaicataldo** [comment]
I can champion this, though I'm currently also championing one more JSCS rule and will get to that one first.

**Comment by kaicataldo** [comment]
@eslint/eslint-team Need some supporters on this!

**Comment by vitorbal** [comment]
:thumbsup: from me!

**Comment by kaicataldo** [comment]
Will be starting on this in the next few days.

**Comment by elodszopos** [comment]
Not to rush anyone, just interested in a status on this. Any news @kaicataldo ?

Would be awesome to have the possibility to enforce stopping the `this.props.blabla` flood.

**Comment by kaicataldo** [comment]
Hey, thanks for checking in - ended up being busier than I expected and haven't gotten to it yet. Finally have some free time this weekend and am hoping to start in on this, but if someone else has the bandwidth please feel free to take it on!

**Comment by valera-rozuvan** [comment]
Hi all! I am really interested in this rule. Any updates? Thanks!

**Comment by ilyavolodin** [comment]
@valera-rozuvan This proposal has been accepted, so it's just waiting for implementation. If you would like to implement this rule, and have questions, @kaicataldo as champion should be able to guide you.

**Comment by valera-rozuvan** [comment]
@kaicataldo Have you seen [alexlafroscia/eslint-plugin-ember-suave](https://github.com/alexlafroscia/eslint-plugin-ember-suave)? That project already has a custom rule for ESLint [ember-suave/prefer-destructuring](https://github.com/alexlafroscia/eslint-plugin-ember-suave/blob/master/docs/rules/prefer-destructuring.md):

Ex:

```
{
  "rules": {
    "ember-suave/prefer-destructuring": ["error", {
      "array": true,
      "object": true
    }]
  }
}
```

Should that implementation be taken as a basis? @alexlafroscia, @hzoo any input from your side?

**Comment by kaicataldo** [comment]
Apologies for not getting around to starting this yet.

@valera-rozuvan Would you like to work on implementing this? It does indeed look like that plugin does what we're trying to achieve here.

**Comment by not-an-aardvark** [comment]
I'm also willing to work on implementing this if no one else is working on it.

**Comment by alexlafroscia** [comment]
@valera-rozuvan AFAIK my implementation there works -- a lot of people in the Ember community use that package and I've never seen issues with it.

**Comment by valera-rozuvan** [comment]
Do we need to dup;icate this functionality in main ESLint? If it's already available as a separate plugin, maybe it's best to spend our effort on something else?

**Comment by kaicataldo** [comment]
@valera-rozuvan Good point. It was already accepted by the team, so it has been deemed fit for core. However, since it's a JSCS compatibility issue, might be worth discussing to see the team thinks. We didn't implement some other rules that exist in other plugins. 

@alexlafroscia If we decide we'd still like this in core, would you be amenable to either making a PR/allowing us to copy over the rule into core (with your author attribution in the rule file)?

**Comment by kaicataldo** [comment]
@eslint/eslint-team ^

**Comment by alexlafroscia** [comment]
@kaicataldo sure, say the word and I'll make the PR.

**Comment by platinumazure** [comment]
@kaicataldo I think we made a promise to JSCS users that they could (eventually) switch to ESLint and get basically all the JSCS functionality. To me it feels like it would be unfair to those users to tell them "actually, you need to use these plugins as well". Makes the conversion harder than it needs to be. I think this should absolutely be ported to core (with proper attribution).

**Comment by kaicataldo** [comment]
@platinumazure I'm with you on that. Just wanted to discuss because there was at least (something regarding jQuery) that we said we would just recommend the plugin for.

**Comment by alexlafroscia** [comment]
I'm putting the PR together to add the rule I already created as part of the core rules.

**Comment by kaicataldo** [comment]
@alexlafroscia Much appreciated!

**Comment by valera-rozuvan** [comment]
Thank you all!
